### ISSUE TITLE
Discussion: priority on environment variable vs compiling time value

### ISSUE THREAD

**Issue by TommyCpp** [open-telemetry/opentelemetry-rust#1225]
Most of our creates generally allow two ways to configure values:

- Use environment variable fetched during runtime
- User provided value during compiling time(`with_xxx`)

However, this is unclear when both of the configurations present which one should take priority.  Regardless we should align on this and use it as a tenet across all our creates.

cc @open-telemetry/rust-approvers

**Comment by djc** [comment]
I find it very confusing to make the compiled-in value take priority over the run-time value (from an environment variable). Code is usually written (and compiled) to function across many environments, whereas the runtime environment is specific to a particular invocation. Clearly the runtime environment, with its better specificity, should be trusted more than the code.

Maybe we should just not provide any way to configure these values in code?

See also this recent discussion in Cargo on a related problem:

https://github.com/rust-lang/cargo/pull/12515

In this case, there is no value provided in code, but there's a clear hierarchy of specificity.

**Comment by TommyCpp** [comment]
I think most of people favors the environment variables. I think we can favor env vars here and make sure it's consistent.

Please leave comments if you disgree @open-telemetry/rust-approvers

**Comment by hdost** [comment]
> I think most of people favors the environment variables. I think we can favor env vars here and make sure it's consistent.
> 
> 
> 
> Please leave comments if you disgree @open-telemetry/rust-approvers 

So while this is not a deviation (more of a clarification), it might also be useful to document. https://github.com/open-telemetry/opentelemetry-rust/issues/1297

**Comment by cijothomas** [comment]
Reopening to further discuss given above.

**Comment by cijothomas** [comment]
This is not consistent enforced everywhere. Adding to Log SDK stable milestone, as this is important to get closure.

**Comment by lalitb** [comment]
Good catch.

I believe the priority order as discussed above should be 
 - Signal specific env-variables
 - Generic env-variables
 - with_xxx compiled values.
 - default values

Quick look at [BatchConfigBuilder](https://github.com/open-telemetry/opentelemetry-rust/blob/7c9447ff6ca43003b00b5ba32a6b0849106b96a9/opentelemetry-sdk/src/logs/log_processor.rs#L742) , the priority order seems to be:
 - with_xxx compiled values
 - env-variables
 - default values

This needs to be fixed. Will raise an issue for this first.

**Comment by cijothomas** [comment]
Thanks. I'd like to revisit the priority order also.

Re-stating my opinion:

if a user does with_xxx in code, then it should be final, and ENV variable should not override that. if user prefer ENV variable based config, then they should not do with_xxx in code and leave it untouched, so they can use ENV variable to configure things.

This was never clarified in spec, so we are left with own policy, but I also want to recheck with Spec to see if there is some incoming guidance.

**Comment by lalitb** [comment]
> if a user does with_xxx in code, then it should be final, and ENV variable should not override that. if user prefer ENV variable based config, then they should not do with_xxx in code and leave it untouched, so they can use ENV variable to configure things.

Typically, environment variables are prioritized over compiled values because they allow runtime configuration without requiring recompilation. I have seen this convention more widely. Do we have a strong reason for otherwise? I can think of - malicious manipulation of environment variables, or someone leaving environment variables set by mistake - but these risks should be mitigated by proper access control and policies.

**Comment by MikeGoldsmith** [comment]
I disagree with code choices being the final decision maker for configuration values. 

As an operator of software, being able to alter behaviour from the outside (eg using environment variables) at runtime is super important. Otherwise development-time choices require software to be modified (eg rebuilt) which is slow & inefficient.

**Comment by cijothomas** [comment]
> I disagree with code choices being the final decision maker for configuration values.
> 
> As an operator of software, being able to alter behaviour from the outside (eg using environment variables) at runtime is super important. Otherwise development-time choices require software to be modified (eg rebuilt) which is slow & inefficient.

Users can decide not to programmatically configure settings, if they want to ensure ENV can be used to determine behavior without redeployment.

Either way, spec has not clarified this behavior so far, leaving this up to individual implementations and is consistent across languages.
https://github.com/open-telemetry/opentelemetry-specification/issues/4413

**Comment by scottgerring** [comment]
> I would've instead expected having to specify some kind of (imaginary) `.with_env_vars()` to enable environment variables being consulted

This is I would say the common pattern and would make it very explicit, but I suppose it's a breaking change. I think this would be the optimal way to do it greenfields.

> If the user explicitly configured something in code (with_endpoint("my-endpoint")), then an ENV variable overriding it can be surprising

I agree with this. The issue is that the call here is so explicit - "use this". Hierarchy of preference works cleanly when its all implicit - "1/ check env_1, 2/ check_file, 3/ check env_var_2" . The sort of thing you'd see with e.g. the AWS SDK credential resolution chain.

[https://12factor.net/config](https://12factor.net/config) encourages the use of environment variables.

**Comment by cijothomas** [comment]
From today's SIG call:

Nahum:
Env is preferred to take over from own experience.
But we need to throw some error level internal-logs when both are specified, so users are aware of what's going on, instead of being taken by surprise.

Anton:
Hardcoded code should be higher priority to avoid being overrides by someone else.

Leaving open for 2 more days and make a final decision and implement it.

**Comment by cijothomas** [comment]
https://github.com/open-telemetry/opentelemetry-specification/issues/4413#issuecomment-2666202644 
It looks like majority of OTel implementations follow the "explicit-code-takes-precedence-over-env" approach. (It is aligned with my proposal).

As discussed in community call today, will leave it open for comments for another 2 days.

**Comment by TommyCpp** [comment]
> explicit-code-takes-precedence-over-env

Languages also supports users to config it to be auto resolved by env var. We can revert the behaivor to be default favor compiling value but I think we should add options like `with_env_vars` to allow code owners to state they allow env var take over. Also makes the migration easier

**Comment by cijothomas** [comment]
> Languages also supports users to config it to be auto resolved by env var.

Rust is also the same right? if user does not explicitly configure in code, then env-var is respected. (Sorry I may not have fully understood this statement!)

> but I think we should add options like with_env_vars to allow code owners to state they allow env var take over

I am not so sure if maintaining additional complexity is worth it. For users who want env_var, they can choose to *not* programmatically configure it. 
Overall, OTel is moving to a totally new configuration model which would become the norm in the future. ENV variable support remains to maintain back-compatibility.

**Comment by cijothomas** [comment]
From community call today:
@TommyCpp 
It is better to offer a `with_env_vars` so users who rely on existing behavior has an easy migration.
Maybe okay to keep it just for 1-2 release and not a permanent setting.


Link to OTel Spec for new declarative config:  https://github.com/open-telemetry/opentelemetry-specification/issues/4374

**Comment by cijothomas** [comment]
First PR done https://github.com/open-telemetry/opentelemetry-rust/pull/2781
More to come.

**Comment by cijothomas** [comment]
> First PR done [#2781](https://github.com/open-telemetry/opentelemetry-rust/pull/2781) More to come.

This is the only Logging SDK specific change required. Removing this issue from Log SDK stable milestone. and moving to Metric SDK milestone.

**Comment by rubberduck203** [comment]
Can someone clarify for me what the _current_ priority is and what it will be after these PRs land?  
I'm currently pretty convinced there was a behavior change between v0.22 and v0.27 that broke me and I'm trying to figure out the best solution in light of this discussion and any upcoming changes.

**Comment by cijothomas** [comment]
> Can someone clarify for me what the _current_ priority is and what it will be after these PRs land? I'm currently pretty convinced there was a behavior change between v0.22 and v0.27 that broke me and I'm trying to figure out the best solution in light of this discussion and any upcoming changes.

it is not consistent today. Some prefer env some does code 😢 
The goal is to :
1. Make it consistent everywhere in the repo
2. Make code take higher priority than env
3. Make the priority order clear in the doc

**Comment by rubberduck203** [comment]
Thank you @cijothomas. That at least tells me how to prepare for the upcoming changes. Much appreciated.
